ddl/index_merge_tmp.go

```diff
@@ -0,0 +1,238 @@
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// IsEnableFastReorg check whether Fast Reorg is allowed.
+func IsEnableFastReorg() bool {
+	return variable.EnableFastReorg.Load()
+}
+
+func (w *mergeIndexWorker) batchCheckTemporaryUniqueKey(txn kv.Transaction, idxRecords []*temporaryIndexRecord) error {
+	idxInfo := w.index.Meta()
+	if !idxInfo.Unique {
+		// non-unique key need no check, just overwrite it,
+		// because in most case, backfilling indices is not exists.
+		return nil
+	}
+
+	batchVals, err := txn.BatchGet(context.Background(), w.originIdxKeys)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// 1. unique-key/primary-key is duplicate and the handle is equal, skip it.
+	// 2. unique-key/primary-key is duplicate and the handle is not equal, return duplicate error.
+	// 3. non-unique-key is duplicate, skip it.
+	for i, key := range w.originIdxKeys {
+		if val, found := batchVals[string(key)]; found {
+			if idxRecords[i].distinct {
+				if !bytes.Equal(val, idxRecords[i].vals) {
+					return kv.ErrKeyExists
+				}
+			}
+			if !idxRecords[i].delete {
+				idxRecords[i].skip = true
+			}
+		} else if idxRecords[i].distinct {
+			// The keys in w.batchCheckKeys also maybe duplicate,
+			// so we need to backfill the not found key into `batchVals` map.
+			batchVals[string(key)] = idxRecords[i].vals
+		}
+	}
+	return nil
+}
+
+// temporaryIndexRecord is the record information of an index.
+type temporaryIndexRecord struct {
+	vals     []byte
+	skip     bool // skip indicates that the index key is already exists, we should not add it.
+	delete   bool
+	unique   bool
+	distinct bool
+}
+
+type mergeIndexWorker struct {
+	*backfillWorker
+
+	index table.Index
+
+	tmpIdxRecords []*temporaryIndexRecord
+	originIdxKeys []kv.Key
+	tmpIdxKeys    []kv.Key
+	jobContext    *JobContext
+}
+
+func newMergeTempIndexWorker(sessCtx sessionctx.Context, id int, t table.PhysicalTable, reorgInfo *reorgInfo, jc *JobContext) *mergeIndexWorker {
+	indexInfo := model.FindIndexInfoByID(t.Meta().Indices, reorgInfo.currElement.ID)
+
+	index := tables.NewIndex(t.GetPhysicalID(), t.Meta(), indexInfo)
+
+	return &mergeIndexWorker{
+		backfillWorker: newBackfillWorker(sessCtx, id, t, reorgInfo, typeAddIndexMergeTmpWorker),
+		index:          index,
+		jobContext:     jc,
+	}
+}
+
+// BackfillDataInTxn merge temp index data in txn.
+func (w *mergeIndexWorker) BackfillDataInTxn(taskRange reorgBackfillTask) (taskCtx backfillTaskContext, errInTxn error) {
+	oprStartTime := time.Now()
+	ctx := kv.WithInternalSourceType(context.Background(), w.jobContext.ddlJobSourceType())
+	errInTxn = kv.RunInNewTxn(ctx, w.sessCtx.GetStore(), true, func(ctx context.Context, txn kv.Transaction) error {
+		taskCtx.addedCount = 0
+		taskCtx.scanCount = 0
```

Comment:
Do we need to update these two fields? Or in the next  PR?

Reply:
Updated.